src/graph.hpp

```diff
@@ -2,15 +2,46 @@
 #define GRAPH_HPP
 
 #include <edge.hpp>
+#include <impl/graph.hpp>
+#include <impl/meta_utils.hpp>
 
 namespace graph {
 
-template<typename Node, typename... Types>
-class Graph {
+class GraphTrait {
+protected:
+    GraphTrait() = default;
+};
+
+class EdgeTrait {
+protected:
+    EdgeTrait() = default;
+};
+
+// TODO: source and sink Node* type.
+// Should we inherit all nodes from base Node class
+// or should we apply more template woodoo?
+class Net : public GraphTrait {
 public:
-    virtual void addEdge(const Edge<Node, Types...>& edge) = 0;
+    Net(int source, int sink) : source(source), sink(sink) {}
+    int source, sink;
 };
 
+class Weighted : public EdgeTrait {
+public:
+    Weighted(uint32_t weight) : weight(weight) {}
+    uint32_t weight;
+};
+
+// TODO: Here I want to apply
+// requires ((std::is_base_of_v<graph::GraphTrait, Ts> || std::is_base_of_v<graph::EdgeTrait, Ts>) && ...)
```

Comment:
Make a single trait that checks whether Ts is derived from GraphTrait nor EdgeTrait:

```
template<typename T>
concept GraphTraitOrEdgeTrait =
    std::is_base_of_v<GraphTrait, T> or
    std::is_base_of_v<GraphTrait, T>;
```

and next:

`requires (GraphTraitOrEdgeTrait<Ts> && ...)`

Reply:
Thanks!